### DISCUSSION TITLE
Database Providers & Installs Using Source

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3621] [General]
The changes in 5.0.1 related to Database Providers have proven to be problematic for developers who are using the "source code" version of Oqtane (ie. either cloning the repo or downloading a ZIP of the source code). There have been reports from 8 different developers related to this problem already. The problem is related to the fact that:

1. Developers do not read documentation/instructions (which have always indicated that the Oqtane.sln should be compiled prior to running the application)
2. They assume they can open the solution and immediately hit F5 (run) and expect that all dependencies are included.
3. The behavior in 5.0.1 is different than it was previously - ie. in versions prior to 5.0.1 developers could clone and run because the database providers were included in the repo as Nuget packages which were auto-installed (however the way they were auto installed could cause .NET upgrade issues in some scenarios). 

#3613 attempts to address this issue by providing a friendly error message:

"The Database Provider {databaseType} Does Not Exist. If This Is A Development Environment Please Ensure You Have Performed A Full Compilation Of All Projects In The Oqtane Solution Prior To Running The Application."

Note that this is not a problem when using the official release packages - it is only a problem when using the source directly from the repo. However it is an important use case.

To be clear, the database provider behavior was changed because users were experiencing upgrade issues transitioning from .NET 7 to .NET 8 - and this problem existed in the official release packages - which is a higher priority scenario than the use case of working directly with source from Github. 

However this ended up being one of those situations where trying to fix a problem uncovered a myriad of other dependencies and behaviors which resulted in a chain reaction of changes to address the new issues as they arose. #3613 is merely the latest attempt to patch a patch. So at this point, although the original .NET upgrade problem was resolved, I am not certain if the current solution is optimal. Now that all scenarios are well known I am wondering if the changes in 5.0.1 should be rolled back and resolved in a different way. 

However this will involve a bunch of work, testing, and churn... so I would like to get some feedback from the community on whether the issues that developers are reporting with using source from Github is significant enough that it needs to be addressed with a completely new solution - or if the current approach is sufficient (ie. with the new message in #3613).

**Comment by W6HBR** [comment]
I would think the new message should be more than sufficient.  Otherwise, you're causing yourself work because others can't follow directions.

**Reply by thabaum** [reply]
I have addressed the issue in documentation as well once those changes get merged.  I can look to do the same anywhere I find in the Oqtane Framework GitHub repo as well.

https://www.oqtane.org/blog/!/24/installing-oqtane-on-iis

This blog is where the directions take people, notice no "Step 5" and "Step 6" are indeed fuzzy here.

This is found as the link directed to getting started from the readme on this project page.

Here is the directions on the documentation with my proposed updates in the Oqtane.Docs GitHub repo:

[Enhances Installation Instructions And Documentation For Latest Oqtane Framework Version 5 With .NET 8](https://github.com/oqtane/oqtane.docs/pull/14)

You can review the changes as I made sure each step had it's own line for clarity including "build".

However I would like to do a 1UP on this and update the link to point from the Blog to documentation and be sure documentation covers everything in the blog merging them together in a very clear way.

I really need what PR's I have submitted reviewed there and added so I can move forward now.  I have done a few different PRs but no responses in that realm.

Good documentation should resolve this issue as it will also include a place for troubleshooting information.

I strongly suggest we review some things and learn to help provide DocFX documentation support.

Once this all has been fully addressed I would see if this is ever an issue again.

**Reply by sbwalker** [reply]
@thabaum I was looking for feedback on whether the current behavior for database providers is sufficient or needs to be refactored. So although it's good document the current behavior - it could ultimately change which would mean the documentation would also need to potentially change or else it will lead to more confusion. 

In regards to documentation, I personally do not have the bandwidth to review all documentation. Similar to the localization repo, other community members need to own the docs repo. Currently I think the owner is @iJungleboy however if he does not have bandwidth we will need to add some additional owners. 

However this raises an important topic. At the Developer Meeting this week we discussed that GitHub may not be the optimal tool/workflow for maintaining docs. Proposals were made for creating alternatives to GitHub. So I do not really understand the sudden focus on updating docs in GitHub?

**Reply by thabaum** [reply]
@sbwalker, My opinion is if all areas can be reviewed to have directions that show everything clearly which can currently be edited and is referenced to build the solution first along with this message it should not be a problem where people post here much. Someone in the community can easily post a link to instructions freeing up time for other things.  A lot of docs created in the source code are read into DocFX.  I found it hard to think it could be unwound from the Oqtane Community Ecosystem.

Revving up the engine on our docs talk! Buckle up for the ride:

  ### Dual Boost:
  We're talking a Wiki pit stop for community vibes and a Docs express lane for the DIYers. Double the power!

  ### Turbocharged DocFX:
  DocFX is our speedster, and folks like @iJungleboy are giving it a high-five. Ready to keep it in the fast lane.

  ### Nitrous Wiki-Docs Blend:
  Let's hit the road with Wiki's community curves and Docs' precision turns. Monthly pit stops for a smooth ride.

  ### Spoiler Alert: Enhancements!
  Tuning up the blog with live writers – that's the turbo boost we need thanks @ADefWebserver for this upcoming contribution. Pedal to the metal, community input style!

  ### New Module: Need for Speed?
  Questions burning rubber about the new module - how do we install it? Updates come with the horsepower, or do we import them? Docs for the racers, you know!

  ### Core Revs, Community Rolls:
  You work the core, we'll cruise the community. No competition, just a turbocharged tandem.

  ### Consistency on the Autobahn:
  Let's keep it steady on the Autobahn. No surprise potholes, especially for our DocFX racers.

  ### Docs Pit Crew for Support:
  Docs are the pit crew – less breakdowns, more open roads for real stuff.

  ### DocFX Drift:
  What’s the DocFX drift? We want to know it's hugging those corners like a pro on the track.

  ### Contributions - Full Throttle!
  Post v5.2, I’m strapping in for contributions. Let’s talk about keeping the docs as fresh as a new coat of wax!

Drop the gears and let’s hit the road!

Cheers!

**Reply by thabaum** [reply]
@sbwalker this [pull request for file read me file change](https://github.com/oqtane/oqtane.framework/pull/3628/files) and others like this is what I am suggesting as I also [suggested in the Oqtane.Docs project](https://github.com/oqtane/oqtane.docs/pull/14).  This should be updated in Blogs or Wiki as well it does not matter.  The key is breaking down the steps instead of combining them into a sentence and adding the importance around building it prior to running it with these instructions.  They are short enough this clarification of what needs to be done first by adding a few steps to count wont hurt.
